# Proyek Analisis Data: [Bike-sharing-dataset]
- **Nama:** [Sultan Juanda Al Muttaqin]
- **Email:** [sultanjuanda123@gmail.com]
- **ID Dicoding:** [MC130D5Y0185]

## Menentukan Pertanyaan Bisnis

- Kapan Waktu Terbaik untuk Menyewa Sepeda?
- Faktor yang Mempengaruhi Penyewaan Sepeda?

## Import Semua Packages/Library yang Digunakan

In [ ]:
import numpy as np
import pandas as pd
import matplotlib
matplotlib.use('Agg')
import matplotlib.pyplot as plt
import seaborn as sns
import os

## Data Wrangling

### Gathering Data

In [ ]:
day_df = pd.read_csv("../var/dataset/day.csv")  # Ganti dengan nama file CSV kamu
day_df.head()  # Menampilkan 5 baris pertama (bisa ganti angkanya)

,instant,dteday,season,yr,mnth,holiday,weekday,workingday,weathersit,temp,atemp,hum,windspeed,casual,registered,cnt
0,1,2011-01-01,1,0,1,0,6,0,2,0.344167,0.363625,0.805833,0.160446,331,654,985
1,2,2011-01-02,1,0,1,0,0,0,2,0.363478,0.353739,0.696087,0.248539,131,670,801
2,3,2011-01-03,1,0,1,0,1,1,1,0.196364,0.189405,0.437273,0.248309,120,1229,1349
3,4,2011-01-04,1,0,1,0,2,1,1,0.200000,0.212122,0.590435,0.160296,108,1454,1562
4,5,2011-01-05,1,0,1,0,3,1,1,0.226957,0.229270,0.436957,0.186900,82,1518,1600


In [ ]:
hour_df = pd.read_csv("../var/dataset/hour.csv")  # Ganti dengan nama file CSV kamu
hour_df.head()  # Menampilkan 5 baris pertama (bisa ganti angkanya)

,instant,dteday,season,yr,mnth,hr,holiday,weekday,workingday,weathersit,temp,atemp,hum,windspeed,casual,registered,cnt
0,1,2011-01-01,1,0,1,0,0,6,0,1,0.24,0.2879,0.81,0.0,3,13,16
1,2,2011-01-01,1,0,1,1,0,6,0,1,0.22,0.2727,0.80,0.0,8,32,40
2,3,2011-01-01,1,0,1,2,0,6,0,1,0.22,0.2727,0.80,0.0,5,27,32
3,4,2011-01-01,1,0,1,3,0,6,0,1,0.24,0.2879,0.75,0.0,3,10,13
4,5,2011-01-01,1,0,1,4,0,6,0,1,0.24,0.2879,0.75,0.0,0,1,1


**Insight:**
- disini kita sudah bisa mengakses dataset dari file csv
- melihat value dari variable

### Assessing Data

In [ ]:
day_df.info()
hour_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 731 entries, 0 to 730
Data columns (total 16 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   instant     731 non-null    int64  
 1   dteday      731 non-null    object 
 2   season      731 non-null    int64  
 3   yr          731 non-null    int64  
 4   mnth        731 non-null    int64  
 5   holiday     731 non-null    int64  
 6   weekday     731 non-null    int64  
 7   workingday  731 non-null    int64  
 8   weathersit  731 non-null    int64  
 9   temp        731 non-null    float64
 10  atemp       731 non-null    float64
 11  hum         731 non-null    float64
 12  windspeed   731 non-null    float64
 13  casual      731 non-null    int64  
 14  registered  731 non-null    int64  
 15  cnt         731 non-null    int64  
dtypes: float64(4), int64(11), object(1)
memory usage: 91.5+ KB
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 17379 entries, 0 to 17378
Data columns (tot

**Insight:**
- disini dapat kita lihat untuk kedua table memiliki tipe data kolom yang kurang tepat yaitu dteday
- nama kolom kedua table kurang jelas atau singkat
- tidak ada data yang missing

In [ ]:
day_df.isna().sum()

,0
instant,0
dteday,0
season,0
yr,0
mnth,0
holiday,0
weekday,0
workingday,0
weathersit,0
temp,0


- tidak ada data yang missing

In [ ]:
print("Jumlah duplikasi: ", day_df.duplicated().sum())

Jumlah duplikasi:  0


- tidak ada data yang duplikat

In [ ]:
day_df.describe()

,instant,season,yr,mnth,holiday,weekday,workingday,weathersit,temp,atemp,hum,windspeed,casual,registered,cnt
count,731.000000,731.000000,731.000000,731.000000,731.000000,731.000000,731.000000,731.000000,731.000000,731.000000,731.000000,731.000000,731.000000,731.000000,731.000000
mean,366.000000,2.496580,0.500684,6.519836,0.028728,2.997264,0.683995,1.395349,0.495385,0.474354,0.627894,0.190486,848.176471,3656.172367,4504.348837
std,211.165812,1.110807,0.500342,3.451913,0.167155,2.004787,0.465233,0.544894,0.183051,0.162961,0.142429,0.077498,686.622488,1560.256377,1937.211452
min,1.000000,1.000000,0.000000,1.000000,0.000000,0.000000,0.000000,1.000000,0.059130,0.079070,0.000000,0.022392,2.000000,20.000000,22.000000
25%,183.500000,2.000000,0.000000,4.000000,0.000000,1.000000,0.000000,1.000000,0.337083,0.337842,0.520000,0.134950,315.500000,2497.000000,3152.000000
50%,366.000000,3.000000,1.000000,7.000000,0.000000,3.000000,1.000000,1.000000,0.498333,0.486733,0.626667,0.180975,713.000000,3662.000000,4548.000000
75%,548.500000,3.000000,1.000000,10.000000,0.000000,5.000000,1.000000,2.000000,0.655417,0.608602,0.730209,0.233214,1096.000000,4776.500000,5956.000000
max,731.000000,4.000000,1.000000,12.000000,1.000000,6.000000,1.000000,3.000000,0.861667,0.840896,0.972500,0.507463,3410.000000,6946.000000,8714.000000


- disini kita melihat ada data yang terlihat seperti inacurate data pada kolom cnt nilai maksimunnya sangat banyak

In [ ]:
day_df[day_df["dteday"].between("2012-09-10", "2012-09-20")][["dteday", "cnt"]]


,dteday,cnt
618,2012-09-10,7525
619,2012-09-11,7767
620,2012-09-12,7870
621,2012-09-13,7804
622,2012-09-14,8009
623,2012-09-15,8714
624,2012-09-16,7333
625,2012-09-17,6869
626,2012-09-18,4073
627,2012-09-19,7591


- setelah kita cek data dari cnt ternyata tidak inacurate ditandai dengan data yang lain juga tidak jauh berbeda

In [ ]:
print("Jumlah duplikasi: ",hour_df.duplicated().sum())
hour_df.describe()

Jumlah duplikasi:  0


,instant,season,yr,mnth,hr,holiday,weekday,workingday,weathersit,temp,atemp,hum,windspeed,casual,registered,cnt
count,17379.0000,17379.000000,17379.000000,17379.000000,17379.000000,17379.000000,17379.000000,17379.000000,17379.000000,17379.000000,17379.000000,17379.000000,17379.000000,17379.000000,17379.000000,17379.000000
mean,8690.0000,2.501640,0.502561,6.537775,11.546752,0.028770,3.003683,0.682721,1.425283,0.496987,0.475775,0.627229,0.190098,35.676218,153.786869,189.463088
std,5017.0295,1.106918,0.500008,3.438776,6.914405,0.167165,2.005771,0.465431,0.639357,0.192556,0.171850,0.192930,0.122340,49.305030,151.357286,181.387599
min,1.0000,1.000000,0.000000,1.000000,0.000000,0.000000,0.000000,0.000000,1.000000,0.020000,0.000000,0.000000,0.000000,0.000000,0.000000,1.000000
25%,4345.5000,2.000000,0.000000,4.000000,6.000000,0.000000,1.000000,0.000000,1.000000,0.340000,0.333300,0.480000,0.104500,4.000000,34.000000,40.000000
50%,8690.0000,3.000000,1.000000,7.000000,12.000000,0.000000,3.000000,1.000000,1.000000,0.500000,0.484800,0.630000,0.194000,17.000000,115.000000,142.000000
75%,13034.5000,3.000000,1.000000,10.000000,18.000000,0.000000,5.000000,1.000000,2.000000,0.660000,0.621200,0.780000,0.253700,48.000000,220.000000,281.000000
max,17379.0000,4.000000,1.000000,12.000000,23.000000,1.000000,6.000000,1.000000,4.000000,1.000000,1.000000,1.000000,0.850700,367.000000,886.000000,977.000000


- disini tidak ada data yang duplikat
- tidak ada data yang inacurate

In [ ]:
hour_df.isna().sum()

,0
instant,0
dteday,0
season,0
yr,0
mnth,0
hr,0
holiday,0
weekday,0
workingday,0
weathersit,0


- tidak ada data yang missing pada dataset hour_df

In [ ]:
datetime_columns = ["dteday"]

for column in datetime_columns:
 day_df[column] = pd.to_datetime(day_df[column])

 day_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 731 entries, 0 to 730
Data columns (total 16 columns):
 #   Column      Non-Null Count  Dtype         
---  ------      --------------  -----         
 0   instant     731 non-null    int64         
 1   dteday      731 non-null    datetime64[ns]
 2   season      731 non-null    int64         
 3   yr          731 non-null    int64         
 4   mnth        731 non-null    int64         
 5   holiday     731 non-null    int64         
 6   weekday     731 non-null    int64         
 7   workingday  731 non-null    int64         
 8   weathersit  731 non-null    int64         
 9   temp        731 non-null    float64       
 10  atemp       731 non-null    float64       
 11  hum         731 non-null    float64       
 12  windspeed   731 non-null    float64       
 13  casual      731 non-null    int64         
 14  registered  731 non-null    int64         
 15  cnt         731 non-null    int64         
dtypes: datetime64[ns](1), floa

- disini kita sudah berhasil mengubah tipe data dari dteday menjadi datetime

### Cleaning Data

In [ ]:
rename_dict = {
    'yr': 'year',
    'mnth': 'month',
    'weekday': 'day_of_week',
    'weathersit': 'weather_situation',
    'windspeed': 'wind_speed',
    'cnt': 'total_rentals',
    'hum': 'humidity'
}
day_df.rename(columns=rename_dict, inplace=True)
hour_df.rename(columns={**rename_dict, 'hr': 'hour'}, inplace=True)
day_df.describe(include="all")

,instant,dteday,season,year,month,holiday,day_of_week,workingday,weather_situation,temp,atemp,humidity,wind_speed,casual,registered,total_rentals
count,731.000000,731,731.000000,731.000000,731.000000,731.000000,731.000000,731.000000,731.000000,731.000000,731.000000,731.000000,731.000000,731.000000,731.000000,731.000000
mean,366.000000,2012-01-01 00:00:00,2.496580,0.500684,6.519836,0.028728,2.997264,0.683995,1.395349,0.495385,0.474354,0.627894,0.190486,848.176471,3656.172367,4504.348837
min,1.000000,2011-01-01 00:00:00,1.000000,0.000000,1.000000,0.000000,0.000000,0.000000,1.000000,0.059130,0.079070,0.000000,0.022392,2.000000,20.000000,22.000000
25%,183.500000,2011-07-02 12:00:00,2.000000,0.000000,4.000000,0.000000,1.000000,0.000000,1.000000,0.337083,0.337842,0.520000,0.134950,315.500000,2497.000000,3152.000000
50%,366.000000,2012-01-01 00:00:00,3.000000,1.000000,7.000000,0.000000,3.000000,1.000000,1.000000,0.498333,0.486733,0.626667,0.180975,713.000000,3662.000000,4548.000000
75%,548.500000,2012-07-01 12:00:00,3.000000,1.000000,10.000000,0.000000,5.000000,1.000000,2.000000,0.655417,0.608602,0.730209,0.233214,1096.000000,4776.500000,5956.000000
max,731.000000,2012-12-31 00:00:00,4.000000,1.000000,12.000000,1.000000,6.000000,1.000000,3.000000,0.861667,0.840896,0.972500,0.507463,3410.000000,6946.000000,8714.000000
std,211.165812,NaN,1.110807,0.500342,3.451913,0.167155,2.004787,0.465233,0.544894,0.183051,0.162961,0.142429,0.077498,686.622488,1560.256377,1937.211452


**Insight:**
- disini saya membuat nama-nama kolomnya menjadi lebih jelas dan tidak disingkat
- Mempermudah proses eksplorasi serta interpretasi data untuk mendapatkan insight lebih jelas

## Exploratory Data Analysis (EDA)

### Explore ...

In [ ]:
hour_df.groupby('hour')['total_rentals'].mean().sort_index()


,total_rentals
hour,
0,53.898072
1,33.375691
2,22.869930
3,11.727403
4,6.352941
5,19.889819
6,76.044138
7,212.064649
8,359.011004


**Insight:**
- Biasanya penyewaan tinggi di jam sibuk (07-09 pagi & 17-19 sore) saat orang pergi/pulang kerja.

In [ ]:
day_df.groupby('day_of_week')['total_rentals'].mean().sort_index()


,total_rentals
day_of_week,
0,4228.828571
1,4338.123810
2,4510.663462
3,4548.538462
4,4667.259615
5,4690.288462
6,4550.542857


- penyewaan terbanyak ada di hari jumat

In [ ]:
day_df.groupby('month')['total_rentals'].mean().sort_index()

,total_rentals
month,
1,2176.338710
2,2655.298246
3,3692.258065
4,4484.900000
5,5349.774194
6,5772.366667
7,5563.677419
8,5664.419355
9,5766.516667


- penyewaan paling banyak pada bulan 9

In [ ]:
day_df.corr()['total_rentals'].sort_values(ascending=False)


,total_rentals
total_rentals,1.000000
registered,0.945517
casual,0.672804
atemp,0.631066
instant,0.628830
dteday,0.628830
temp,0.627494
year,0.566710
season,0.406100
month,0.279977


- Cuaca & suhu adalah faktor utama yang mempengaruhi jumlah
  penyewaan.
- Pengguna terdaftar mendominasi penyewaan sepeda
  dibandingkan pengguna kasual.
- Penyewaan lebih tinggi saat musim tertentu dan semakin
  meningkat setiap tahunnya.
- Faktor eksternal seperti angin kencang, hujan, dan hari
  libur menurunkan jumlah penyewaan.
  

In [ ]:
day_df.groupby('weather_situation')['total_rentals'].mean()


,total_rentals
weather_situation,
1,4876.786177
2,4035.862348
3,1803.285714


- Semakin buruk cuaca, semakin rendah jumlah penyewaan sepeda.

In [ ]:
print("Korelasi humidity dengan total_rentals:")
print(day_df[['humidity', 'total_rentals']].corr())

print("\nKorelasi wind_speed dengan total_rentals:")
print(day_df[['wind_speed', 'total_rentals']].corr())


Korelasi humidity dengan total_rentals:
               humidity  total_rentals
humidity       1.000000      -0.100659
total_rentals -0.100659       1.000000

Korelasi wind_speed dengan total_rentals:
               wind_speed  total_rentals
wind_speed       1.000000      -0.234545
total_rentals   -0.234545       1.000000


- Ini berarti semakin tinggi kelembaban udara, semakin sedikit jumlah penyewaan sepeda.
- Semakin tinggi kecepatan angin, semakin rendah jumlah penyewaan sepeda.

## Visualization & Explanatory Analysis

### Pertanyaan 1:  Kapan Waktu Terbaik untuk Menyewa Sepeda?


In [ ]:
# Hitung rata-rata penyewaan per jam
hourly_rentals = hour_df.groupby('hour')['total_rentals'].mean()

plt.figure(figsize=(10, 5))
sns.lineplot(x=hourly_rentals.index, y=hourly_rentals.values, marker='o')
plt.title("Rata-rata Penyewaan Sepeda per Jam")
plt.xlabel("Jam")
plt.ylabel("Rata-rata Penyewaan")
plt.grid(True)
plt.show()  # Pastikan hanya ada satu plt.show()

### Pertanyaan 2: Faktor yang Mempengaruhi Penyewaan Sepeda?

In [ ]:
corr_matrix = day_df[['temp', 'humidity', 'wind_speed', 'total_rentals']].corr()

# Heatmap korelasi
plt.figure(figsize=(8, 5))
sns.heatmap(corr_matrix, annot=True, cmap='coolwarm', linewidths=0.5)
plt.title("Heatmap Korelasi Faktor terhadap Jumlah Penyewaan Sepeda")
plt.show()

**Insight:**
- waktu terbaik untuk menyewa sepeda adalah saat jam pagi
  hari (sekitar jam 7-9) dan sore/malam (sekitar jam 17-19)
- Suhu (temperature) memiliki korelasi positif tinggi dengan
  jumlah penyewaan, artinya semakin hangat cuaca, semakin banyak orang menyewa sepeda.

## Analisis Lanjutan (Opsional)

## Conclusion

- Waktu terbaik untuk menyewa sepeda adalah pukul 08:00 dan
  17:00 - 18:00, bertepatan dengan jam berangkat dan pulang kerja. Penyewaan menurun drastis pada malam hari.
- Faktor utama yang mempengaruhi penyewaan adalah suhu
  (positif 0.63), sementara kelembaban dan kecepatan angin berdampak negatif meskipun lemah. Cuaca buruk juga menurunkan jumlah penyewaan.